### DISCUSSION TITLE
Allow using both `enabled` and `skipToken` in the same query

### DISCUSSION THREAD

**Discussion by jennings** [TanStack/query#8222] [General] [answered]
I've read #7057 and it still seems to me that `skipToken` should be able to override `enabled: true`. The reply doesn't really explain why that behavior is either not possible or undesirable.

We have the same use case that @averageokguy had: One file defines the queries using  `queryOptions`. In this file, we use `skipToken` to determine when a query _can possibly_ run:

```
// queries.ts
const postByAuthorId = (authorId: number | undefined) =>
  queryOptions({
    queryKey: ["posts-by-author", authorId],
    queryFn: authorId ? () => fetchPostByAuthorId(authorId) : skipToken,
  });
```

In our components, we may have additional restrictions on when we want the query to run. For example, maybe we have a "Show more by this author" that shows more of this author's posts. I'd _like_ to handle it by using `enabled`:

```
function Post(props) {
  const [showMoreByAuthor, setShowMoreByAuthor] = useState(false);

  const { data } = useQuery({
    ...postByAuthorId(props.post.authorId),
    enabled: showMoreByAuthor,
  });

  return <div>
    ...
    {showMoreByAuthor ?
        <PostList posts={data} />
      : <button onClick={() => setShowMoreByAuthor(true)}>Show more</button>}
  </div>
}
```

My intention here is that the query _cannot_ run if `queryFn` resolves to `skipToken`. It would only run if _both_ `queryFn` is a function _and_ `enabled` is true/undefined.

I think throwing on `refetch` is reasonable if the `queryFn` is `skipToken`, but it seems like `{ queryFn: skipToken, enabled: true }` should be allowed and handled as a disabled query.

**Comment by TkDodo** [comment]
`skipToken` is implemented in a way that it sets enabled to false:

https://github.com/TanStack/query/blob/edec4a64c0ca14f407584485c9b1bc4d730958ee/packages/query-core/src/queryClient.ts#L598-L603

> but it seems like { queryFn: skipToken, enabled: true } should be allowed and handled as a disabled query.

If I understand this correctly, we could achieve this by simplifying the condition?

```
if (defaultedOptions.queryFn === skipToken) {
  defaultedOptions.enabled = false
}
```

?

**Reply by jennings** [reply]
That seems right to me!

**Reply by TkDodo** [reply, marked as answer]
see: https://github.com/TanStack/query/pull/8407
